### PR TITLE
Clean up Logging

### PR DESCRIPTION
**Pull request by diesieben07**
Like discussed on IRC.

- FMLRelaunchLog is deprecated all FML logging goes through FMLLog now.
- The methods in FMLLog are deprecated, as they do not offer any additional value on top of log4j.
- All logging uses the log4j logger (`FMLLog.log`) directly.
- All reliance on `String.format` has been removed and replaced with log4j's default formatting using `"{}"` as the placeholder.

This fixes https://github.com/MinecraftForge/MinecraftForge/issues/3713 (and maybe others).

### CODE REVIEW

File: src/main/java/net/minecraftforge/fml/common/FMLLog.java [hunk 1]
```diff
@@ -19,73 +19,107 @@
...
+    static {
+        log = LogManager.getLogger("FML");
+        // Default side to client for test harness purposes
+        Side side = FMLLaunchHandler.side();
+        if (side == null) side = Side.CLIENT;
+        ThreadContext.put("side", side.name().toLowerCase(Locale.ENGLISH));
+
+        log.debug("Injecting tracing printstreams for STDOUT/STDERR.");
+        System.setOut(new TracingPrintStream(LogManager.getLogger("STDOUT"), System.out));
+        System.setErr(new TracingPrintStream(LogManager.getLogger("STDERR"), System.err));
```
Comment (liach) [inline src/main/java/net/minecraftforge/fml/common/FMLLog.java:51]: Really, trace for errors?

Comment (kashike) [reply src/main/java/net/minecraftforge/fml/common/FMLLog.java:51]: https://github.com/MinecraftForge/MinecraftForge/blob/1.11.x/src/main/java/net/minecraftforge/fml/common/TracingPrintStream.java#L41-L60

Comment (liach) [reply src/main/java/net/minecraftforge/fml/common/FMLLog.java:51]: Still doubt if stderr should be error or info.

**Comment by mezz** [comment]
Logs with bad arguments:
 * `B3DModel`
   * `if(oldKey.getPos() != null) logger.error("Duplicate keys: %s and %s (ignored)", oldKey, key);`
   * `if(oldKey.getScale() != null) logger.error("Duplicate keys: %s and %s (ignored)", oldKey, key);`
   * `if(oldKey.getRot() != null) logger.error("Duplicate keys: %s and %s (ignored)", oldKey, key);`
 * `EntitySpawnHandler` 
   * `FMLLog.log.error("A severe problem occurred during the spawning of an entity at ({}, {}, {})", spawnMsg.rawZ, e);`
 * `FMLLaunchHandler`
   * `FMLLog.log.info("Forge Mod Loader version {}.{}.{}.{} for Minecraft {} loading", FMLInjectionData.major, FMLInjectionData.minor, FMLInjectionData.rev, FMLInjectionData.build, FMLInjectionData.mccversion, FMLInjectionData.mcpversion);`
 * `FMLSanityChecker`
   * `log.error("Technical information: ClientBrandRetriever was at %s, there were {} certificates for it", codeSource.getLocation(), certCount);`
 * `ModAPITransformer`
   * `if (logDebugInfo) FMLLog.log.debug("Optional removal - interface {} removed from type signature");`
 * `ModContainerFactory`
   * `FMLLog.log.error("Critical error : cannot register mod container type {}, it has an invalid constructor", e);`
 * `ModelLoaderRegistryDebug`
   * `logger.info("%b", shouldIncrement);`
 * `TextureDump`
   * `logger.info("Exported png to: %s", output.getAbsolutePath());`

Possible NPE:
 * `TileEntity`
   * `net.minecraftforge.fml.common.FMLLog.log.error("A TileEntity {}({}) has thrown an exception during loading, its state cannot be restored. Report this to the mod author",
                    s, oclass.getName(), throwable1);` 
     * `oclass` may be null there.

**Comment by mezz** [comment]
String concatenation:
 * `ConfigTest`
   * `logger.debug("Old: " + CONFIG_TYPES.bool);`
   * `logger.debug("New: " + CONFIG_TYPES.bool);`
   * `logger.debug("After sync: " + CONFIG_TYPES.bool);`
 * `DifficultyChangeEventTest`
   * `logger.info("Difficulty changed from " + event.getOldDifficulty() + " to " + event.getDifficulty());`
* `DynBucketTest`
  * `logger.info("Expected 1: " + handler.getStackInSlot(i));`
  * `logger.info("Expected 2: " + handler2.getStackInSlot(i));`
  * `logger.info("Joined: " + joined.getStackInSlot(i));`
* `EnchantmentLevelSetTest`
  * `logger.info("Enchantment row: " + event.getEnchantRow() + ", level: " + event.getLevel());`
* `EntityTravelToDimensionEventTest`
  * `logger.info("Travelling to Dimension " + event.getDimension() + " Entity: " + event.getEntity());`
* `EquipmentChangeTest`
  * `logger.info("[Equipment-Change] " + event.getEntity() + " changed his Equipment in " + event.getSlot() + " from " + event.getFrom() + " to " + event.getTo());`
* `FluidHandlerTest`
  * `logger.info("Draining " + stackString(stack) + " gives " + fluidString(drain) + " and " + stackString(drainedStack));`
  * `logger.info("Filling " + stackString(stack) + " with " + fluidString(new FluidStack(fluid, filled)) + " gives " + stackString(filledStack));`
* `ForgeModContainer`
  * `log.debug("\t" + name);`
* `GradleForgeHacks`
  * `GradleStartCommon.LOGGER.info("Found and added coremod: " + clazz);`
* `HandshakeMessageHandler`
  * `FMLLog.log.debug(stateType.getSimpleName() + ": " + msg.toString(stateType) + "->" + state.getClass().getName().substring(state.getClass().getName().lastIndexOf('.')+1)+":"+state);`
  * `FMLLog.log.debug("  Next: " + newState.name());`
  * `FMLLog.log.debug(stateType.getSimpleName() + ": null->" + state.getClass().getName().substring(state.getClass().getName().lastIndexOf('.')+1)+":"+state);`
  * `FMLLog.log.debug("  Next: " + newState.name());`
* `ItemFishedTest`
  * `logger.info("Item fished in Biome " + biomeFishedAt.getBiomeName());`
* `MinecraftForge`
  * `ForgeModContainer.log.debug("\t" + name);`
* `ModelAnimationDebug`
  * `logger.info("Event: " + event.event() + " " + event.offset() + " " + getPos() + " " + time);`
* `ModelLoaderRegistryDebug`
  * `logger.info("click " + counter);`
* `NeighborNotifyEventTest`
  * `logger.info(event.getPos().toString() + " with face information: " + event.getNotifiedSides());`
* `PermissionAPI`
  * `FMLLog.log.warn("Replacing " + permissionHandler.getClass().getName() + " with " + handler.getClass().getName());`
* `PlayerSetSpawnTest`
  * `logger.info(event.isForced() + " " + event.getNewSpawn().toString());`
* `TestCapabilityMod`
  * `logger.info("Hi I'm a " + inv.getOwnerType());`
  * `logger.info("Test value is now: " + (cap.getVal() ? "TRUE" : "FALSE"));`
* `TileEntityLoadingTest`
  * `logger.info("World: " + world + ", Pos: " + pos + ", State: " + world.getBlockState(pos));`
* `TracingPrintStream`
  * `logger.info(getPrefix() + o);`
  * `logger.info(getPrefix() + s);`
* `Yggdrasil`
  * `LogManager.getLogger("FMLTWEAK").error("-- Login failed!  " + e.getMessage());`

**Comment by diesieben07** [comment]
Thank you for the detailed report, I have addressed the missing pieces.

**Comment by bs2609** [comment]
@mezz, is this actually intended for 1.11 or 1.12?

**Comment by mezz** [comment]
@bs2609 This is intended for 1.11 if possible.

**Comment by mezz** [comment]
Checked things over again and it looks good to me.

File: src/main/java/net/minecraftforge/fml/common/FMLModContainer.java [hunk 2]
```diff
@@ -510,7 +516,7 @@ private void parseSimpleFieldAnnotation(SetMultimap<String, ASMData> annotations
                 catch (Exception e)
                 {
                     Throwables.propagateIfPossible(e);
-                    FMLLog.log(getModId(), Level.WARN, e, "Attempting to load @%s in class %s for %s and failing", annotationName, targets.getClassName(), mc.getModId());
+                    modLog.warn(modLog.getMessageFactory().newMessage("Attempting to load @%s in class %s for %s and failing", annotationName, targets.getClassName(), mc.getModId()), e);
```
Comment (mezz) [inline src/main/java/net/minecraftforge/fml/common/FMLModContainer.java]: Please fix the `%s`s here

**Review by mezz — CHANGES_REQUESTED** [review]

**Comment by diesieben07** [comment]
Fixed. What about 1.12.x?

**Comment by mezz** [comment]
I can apply this to 1.12 once this is merged.

**Review by mezz — APPROVED** [review]

**Comment by mezz** [comment]
Merged and applied to 1.12 as well.

**Comment by bs2609** [comment]
All the merge conflicts ;_;

**Pull request merged**